restirct max size of http header and user metadata

aead opened the pull request:
## Description

S3 only allows http headers with a size of 8 KB and user-defined metadata
with a size of 2 KB. This change adds a new API error and returns this
error to clients which sends to large http requests.

## Motivation and Context
Fixes #4634 

## How Has This Been Tested?
Add test cases for PutObjectHandler

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

krisis commented on cmd/api-errors.go:
```diff
@@ -561,6 +562,11 @@ var errorCodeResponse = map[APIErrorCode]APIError{
 	},
+	ErrMetadataTooLarge: {
+		Code:           "InvalidArgument",
+		Description:    "Your metadata headers exceed the maximum allowed metadata size.",
```
If we intend to have the string representation of `errMetadataTooLarge` same as `ErrMetadataTooLarge`'s description, then we should do the following,
```go
ErrMetadataTooLarge: {
	Code:           "InvalidArgument",
	Description:    errMetadataTooLarge.String(),
	HTTPStatusCode: http.StatusBadRequest,
},
```

harshavardhana commented on cmd/handler-utils.go:
```diff
@@ -114,16 +114,24 @@ func extractMetadataFromHeader(header http.Header) (map[string]string, error) {
 			metadata[supportedHeader] = header.Get(supportedHeader)
 		}
 	}
+	const MaxHeaderSize = 8 * 1024
```
We should define the constants here.. ideally these should be global. With more descriptive naming 

```
putObjectMaxHeaderSize   = 8 * humanize.KiByte
putObjectMaxUserDataSize = 2 * humanize.KiByte
```

harshavardhana replied:
A relevant AWS S3 docs reference URL should also be commented to indicate this spec restriction.

krishnasrinivas reviewed (commented):
@aead as discussed in person header size restrictions applies to all the calls, hence headerSanityCheck() should be done for all the calls

balamurugana reviewed (commented):
is it good idea to use `MaxHeaderBytes` of HTTP server?  Ref: https://golang.org/pkg/net/http/#Server

Commit:
restirct max size of http header and user metadata (#4634)

S3 only allows http headers with a size of 8 KB and user-defined metadata
with a size of 2 KB. This change adds a new API error and returns this
error to clients which sends to large http requests.

Fixes #4634

krishnasrinivas commented on cmd/generic-handlers.go:
```diff
@@ -65,6 +65,48 @@ func (h requestSizeLimitHandler) ServeHTTP(w http.ResponseWriter, r *http.Reques
+	h.handler.ServeHTTP(w, r)
+}
+
+func (h requestHeaderSizeLimitHandler) checkHTTPHeaderSize(header http.Header) error {
```
checkHTTPHeaderSize need not be a method on requestHeaderSizeLimitHandler, it can be an independent func.
If you rename it to isHTTPHeaderSizeTooLarge, then it can just return `bool` and you dont need to introduce the error errMetadataTooLarge

krishnasrinivas commented on cmd/generic-handlers.go:
```diff
@@ -65,6 +65,44 @@ func (h requestSizeLimitHandler) ServeHTTP(w http.ResponseWriter, r *http.Reques
+	h.Handler.ServeHTTP(w, r)
+}
+
+func isHTTPHeaderSizeTooLarge(header http.Header) bool {
```
Should have pointed this out earlier, this function needs unit tests

aead replied:
done

harshavardhana commented on cmd/generic-handlers.go:
```diff
@@ -62,6 +62,44 @@ func (h requestSizeLimitHandler) ServeHTTP(w http.ResponseWriter, r *http.Reques
 }
 
+const (
+	maxHeaderSize   = 8 * 1024
```
Write comments and point to the relevant AWS S3 documentation.

aead replied:
done

harshavardhana commented on cmd/generic-handlers.go:
```diff
@@ -62,6 +62,44 @@ func (h requestSizeLimitHandler) ServeHTTP(w http.ResponseWriter, r *http.Reques
+	return requestHeaderSizeLimitHandler{h}
+}
+func (h requestHeaderSizeLimitHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if isHTTPHeaderSizeTooLarge(r.Header) {
```
Comments..

aead replied:
done

harshavardhana commented on cmd/handler-utils.go:
```diff
@@ -97,6 +97,11 @@ func path2BucketAndObject(path string) (bucket, object string) {
 	return bucket, object
 }
 
+var userMetadataKeyPrefixes = []string{
```
A comment.. would be necessary for a global variable. If this is not used anywhere define them inside the extractMetadataFromHeader() .

aead replied:
done
